lib/declare_schema/model/field_spec.rb

```diff
@@ -110,35 +129,73 @@ def same_type?(col_spec)
-                  cast_type.deserialize(col_spec.default) != cast_type.deserialize(default)
-                end
-              else
-                col_value = col_spec.send(k)
-                if col_value.nil? && native_type
-                  col_value = native_type[k]
-                end
-                col_value != send(k)
-              end
+        !same_as(col_spec)
+      end
+
+      def same_as(col_spec)
+        same_type?(col_spec) &&
+          same_attributes?(col_spec) &&
+          same_charset_and_collation?(col_spec)
+      end
+
+      private
+
+      def same_attributes?(col_spec)
+        native_type = native_types[type]
+        check_attributes = [:null, :default]
+        check_attributes += [:precision, :scale] if sql_type == :decimal && !col_spec.is_a?(SQLITE_COLUMN_CLASS)  # remove when rails fixes https://rails.lighthouseapp.com/projects/8994-ruby-on-rails/tickets/2872
+        check_attributes -= [:default] if sql_type == :text && col_spec.class.name =~ /mysql/i
+        check_attributes << :limit if sql_type.in?([:string, :binary, :varbinary, :integer, :enum]) ||
+                                      (sql_type == :text && self.class.mysql_text_limits?)
+        check_attributes.all? do |k|
+          if k == :default
+            case Rails::VERSION::MAJOR
+            when 4
+              col_spec.type_cast_from_database(col_spec.default) == col_spec.type_cast_from_database(default)
+            else
+              cast_type = ActiveRecord::Base.connection.lookup_cast_type_from_column(col_spec) or raise "cast_type not found for #{col_spec.inspect}"
+              cast_type.deserialize(col_spec.default) == cast_type.deserialize(default)
+            end
+          else
+            col_value = col_spec.send(k)
+            if col_value.nil? && native_type
+              col_value = native_type[k]
             end
+            col_value == send(k)
           end
+        end
       end
 
-      private
+      def same_charset_and_collation?(col_spec)
+        if type.in?([:text, :string])
+          current_settings = collation_and_charset_for_column(col_spec)
+          current_settings[:collation] == collation
+        else
+          true
+        end
+      end
+
+      def collation_and_charset_for_column(col_spec)
+        column_name   = col_spec.name
+        table_name    = col_spec.table_name
+
+        if ActiveRecord::Base.connection.class.name.match?(/mysql/i)
+          database_name = ActiveRecord::Base.connection.current_database
+
+          query = <<~EOS
+            SELECT C.character_set_name, C.collation_name
+            FROM information_schema.`COLUMNS` C
+            WHERE C.table_schema = "#{database_name}" AND C.table_name = "#{table_name}" AND C.column_name = "#{column_name}";
+          EOS
+          defaults = ActiveRecord::Base.connection.select_one(query)
```

Comment:
It looks like `query` is just used once. If so, why not just put it inline?
```
defaults = ActiveRecord::Base.connection.select_one(<<~EOS)
            SELECT C.character_set_name, C.collation_name
            FROM information_schema.`COLUMNS` C
            WHERE C.table_schema = "#{database_name}" AND C.table_name = "#{table_name}" AND C.column_name = "#{column_name}";
          EOS
```

Reply:
Also, the connection has methods to quote the name of a table and column and I think schema. Best to use those rather than do interpolation inside quotes. In fact it's a bug if any of those names were to have a character in it that needs to be escaped.

Reply:
Done 851956a5c5a4c55ab99e69a9b461201df9c7251d